File: src/VirtualMachine.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

using System;

namespace peak {
	using interpreter;
	public class VirtualMachine {
		public static ModulePool modulePool { get => ModulePool.instance; }
		public static VirtualJourney Load(string src) {
			var executer = Executer.Create(src);
			if (executer != null) {
				return new VirtualJourney(executer);
			}
			return null;
		}
		public static VirtualJourney LoadFile(string filename) {
			var src = VirtualTool.OpenSrc(filename);
			return Load(src);
		}
		public static void LocateLogger(Action<string> logger) {
			LocateLogger(logger, logger);
		}
		public static void LocateLogger(Action<string> echo, Action<string> errorLogger) {
			System0.LocateEcho(echo);
			ErrorLogger.Locate(errorLogger);
		}
		public static void LocateOpenSrc(Func<string, string> func) {
			System0.LocateOpenSrc(func);
		}
	}
} // namespace peak

File: src/virtual_machine.cpp

#include "virtual_machine.h"
#include "base/error_logger.h"
#include "runtime/executer.h"
#include "runtime/module_pool.h"
#include "runtime/system.h"
#include "virtual_journey.h"

using namespace peak;

std::shared_ptr<VirtualJourney> VirtualMachine::Load(const std::string& src) {
	auto executer = Executer::Create(src);
	if (executer) {
		return std::make_shared<VirtualJourney>(executer);
	}
	return nullptr;
}

std::shared_ptr<VirtualJourney> VirtualMachine::LoadFile(const std::string& filename) {
	std::string absPath;
	return Load(System::OpenSrc(filename, absPath));
}

void VirtualMachine::LocateLogger(std::function<void(const std::string&)> logger) {
	LocateLogger(logger, logger);
}
void VirtualMachine::LocateLogger(std::function<void(const std::string&)> echo, std::function<void(const std::string&)> errorLogger) {
	System::LocateEcho(echo);
	ErrorLogger::Locate(errorLogger);
}

void VirtualMachine::LocateLoadText(std::function<std::string(const std::string&)> func) {
	System::LocateLoadText(func);
}

ModulePool* VirtualMachine::GetModulePool() {
	return ModulePool::GetInstance();
}
